ydb/core/tx/datashard/prefix_kmeans.cpp

```diff
@@ -494,28 +470,100 @@ class TPrefixKMeansScan final: public TPrefixKMeansScanBase, private TCalculatio
             cells[0] = TCell::Make(Parent);
             TSerializedCellVec::UnsafeAppendCells(cells, pk);
             PrefixBuf.AddRow(TSerializedCellVec{std::move(pk)}, TSerializedCellVec::Serialize({}));
-        } else if (!TCellVectorsEquals{}(Prefix.GetCells(), key.subspan(0, PrefixColumns))) {
-            HasNextPrefix = true;
-            return EScan::Reset;
         }
-        ++ReadRows;
-        ReadBytes += CountBytes(key, row);
-        switch (State) {
-            case EState::SAMPLE:
-                return FeedSample(row);
-            case EState::KMEANS:
-                return FeedKMeans(row);
-            case EState::UPLOAD_BUILD_TO_BUILD:
-                return FeedUploadBuild2Build(key, row);
-            case EState::UPLOAD_BUILD_TO_POSTING:
-                return FeedUploadBuild2Posting(key, row);
-            default:
-                Y_ASSERT(false);
-                return EScan::Final;
+
+        if (IsFirstPrefixFeed && IsPrefixRowsValid) {
+            PrefixRows.AddRow(TSerializedCellVec{key}, TSerializedCellVec::Serialize(*row));
+            if (PrefixRows.IsReachLimits(Limits)) {
+                PrefixRows.Clear();
+                IsPrefixRowsValid = false;
+            }
+        }
+
+        Feed(key, *row);
+
+        return ShouldWaitUpload() ? EScan::Sleep : EScan::Feed;
+    }
+
+    EScan Exhausted() final
+    {
+        LOG_D("Exhausted " << Debug());
+
+        if (!FinishPrefix()) {
+            return EScan::Reset;
         }
+            
+        IsExhausted = true;
+        
+        // call Seek to wait uploads
+        return EScan::Reset;
     }
 
 private:
+    bool FinishPrefix()
+    {
+        if (FinishPrefixImpl()) {
+            StartNewPrefix();
+            LOG_D("FinishPrefix finished " << Debug());
+            return true;
+        } else {
+            IsFirstPrefixFeed = false;
+            
+            if (IsPrefixRowsValid) {
+                LOG_D("FinishPrefix not finished, manually feeding " << PrefixRows.Size() << " saved rows " << Debug());
+                for (ui64 iteration = 0; ; iteration++) {
+                    for (const auto& [key, row_] : *PrefixRows.GetRowsData()) {
+                        TSerializedCellVec row(row_);
+                        Feed(key.GetCells(), row.GetCells());
+                    }
+                    if (FinishPrefixImpl()) {
+                        StartNewPrefix();
+                        LOG_D("FinishPrefix finished in " << iteration << " iterations " << Debug());
+                        return true;
+                    } else {
+                        LOG_D("FinishPrefix not finished in " << iteration << " iterations " << Debug());
+                    }
+                }
+            } else {
+                LOG_D("FinishPrefix not finished, rescanning rows " << Debug());
```

Comment:
Not sure, will we read smaller?
```suggestion
                LOG_D("FinishPrefix not finished, rescanning rows " << Debug());
                if (State == Upload) {
                   lead.SetTags(ScanTags);
                } else {
                   lead.SetTags({&EmbeddingTag, 1});
                   EmbeddingPosForKMeans = 0;
                }
```

Reply:
We may, but as the first pass require to read all needed columns that seems not much important

Reply:
>  but as the first pass require to read all needed columns that seems not much important

But count of other passes is more than 1 or 2, it even can be configurable see number of rounds

Reply:
We can't read only embedding because when we get next prefix row we expect it to have all of them

Also reading only some tags help only is scheme has different column families

Prefer to leave it as is

Reply:
> We can't read only embedding because when we get next prefix row we expect it to have all of them

We can for kmeans recompute steps

But ok